### PR TITLE
Feature/configuration

### PR DESCRIPTION
**Pull request by felipemonzon**
new enpoint is added to return configuration
- GET /config
- Return site-key to front end

### CODE REVIEW

File: lib/api/shared/src/main/scala/net/wiringbits/api/ApiClient.scala [hunk 1]
```diff
@@ -11,19 +11,28 @@ import scala.util.{Failure, Success, Try}
...
+
   def updatePassword(jwt: String, request: UpdatePassword.Request): Future[UpdatePassword.Response]
+
   def getUserLogs(jwt: String): Future[GetUserLogs.Response]
 
   def adminGetUserLogs(userId: UUID): Future[AdminGetUserLogs.Response]
+
   def adminGetUsers(): Future[AdminGetUsers.Response]
+
+  def getConfig(): Future[Config.Response]
```
Comment (caiseL) [inline lib/api/shared/src/main/scala/net/wiringbits/api/ApiClient.scala:35]: I'm not very sure about the name. Using `getConfig`/`Config.Response` doesn't sound very specific. I'll suggest you change it to make it clear that we're talking about getting frontend config `getFrontendConfig`, `getWebConfig`, `getWebAppConfig` or something else that sounds more explicit

File: server/src/main/scala/controllers/ConfigController.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+class ConfigController @Inject() (
+    configAction: ConfigAction
+)(implicit cc: ControllerComponents, ec: ExecutionContext)
+    extends AbstractController(cc) {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  def getConfig() = handleGET { _ =>
+    logger.info("get configs")
+    for {
+      response <- configAction.apply()
```
Comment (caiseL) [inline server/src/main/scala/controllers/ConfigController.scala:20]: You can skip this apply call and instead to `configAction()`

File: server/src/main/scala/controllers/ConfigController.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import scala.concurrent.ExecutionContext
+
+class ConfigController @Inject() (
+    configAction: ConfigAction
+)(implicit cc: ControllerComponents, ec: ExecutionContext)
+    extends AbstractController(cc) {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  def getConfig() = handleGET { _ =>
+    logger.info("get configs")
```
Comment (caiseL) [inline server/src/main/scala/controllers/ConfigController.scala:18]: `Get configs`/`Get frontend config`, etc

File: server/src/test/scala/controllers/ConfigControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  when(clock.instant()).thenReturn(Instant.now())
+
+  override def guiceApplicationBuilder(container: PostgreSQLContainer): GuiceApplicationBuilder =
+    super
+      .guiceApplicationBuilder(container)
+      .overrides(
+        inject.bind[Clock].to(clock)
+      )
+
+  private def createHMACToken(token: UUID): String = {
```
Comment (caiseL) [inline server/src/test/scala/controllers/ConfigControllerSpec.scala:36]: Do you need this function?

File: server/src/test/scala/controllers/ConfigControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import utils.LoginUtils
+
+import java.time.temporal.ChronoUnit
+import java.time.{Clock, Instant}
+import java.util.UUID
+import scala.concurrent.Future
+
+class ConfigControllerSpec extends PlayPostgresSpec with LoginUtils {
+
+  private val clock = mock[Clock]
```
Comment (caiseL) [inline server/src/test/scala/controllers/ConfigControllerSpec.scala:26]: Is the clock mock necessary?

File: server/src/test/scala/controllers/ConfigControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      .guiceApplicationBuilder(container)
+      .overrides(
+        inject.bind[Clock].to(clock)
+      )
+
+  private def createHMACToken(token: UUID): String = {
+    TokensHelper.doHMACSHA1(token.toString.getBytes, app.injector.instanceOf[UserTokensConfig].hmacSecret)
+  }
+
+  "POST /config" should {
```
Comment (caiseL) [inline server/src/test/scala/controllers/ConfigControllerSpec.scala:40]: `GET /config`

**Review by caiseL — COMMENTED** [review]
Great progress! I think the next step based on #187  is to update the frontend to load these settings

File: lib/api/shared/src/main/scala/net/wiringbits/api/ApiClient.scala (hunk 1 above)
Comment (AlexITC) [reply lib/api/shared/src/main/scala/net/wiringbits/api/ApiClient.scala:35]: I'd suggest `def getEnvironmentConfig: Future[GetEnvironmentConfig.Response]`

File: lib/api/shared/src/main/scala/net/wiringbits/api/models/Config.scala [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package net.wiringbits.api.models
+
+import play.api.libs.json.{Format, Json}
+
+object Config {
+  case class Response(siteKey: String)
```
Comment (AlexITC) [inline lib/api/shared/src/main/scala/net/wiringbits/api/models/Config.scala:6]: What's the siteKey for? I'd name it `recaptchaSiteKey`

File: server/src/main/resources/routes [hunk 8]
```diff
@@ -50,3 +50,6 @@ PUT /admin/tables/:tableName/:ID net.wiringbits.webapp.utils.admin.controllers.A
 
 # delete table resource
 DELETE /admin/tables/:tableName/:ID net.wiringbits.webapp.utils.admin.controllers.AdminController.delete(tableName, ID)
+
+# get configs
+GET /config controllers.ConfigController.getConfig()
```
Comment (AlexITC) [inline server/src/main/resources/routes:55]: Let's change this to `GET /environment-config controllers.EnvironmentConfigController.getEnvironmentConfig`

File: server/src/main/scala/net/wiringbits/actions/ConfigAction.scala [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class ConfigAction @Inject() (
+    reCaptchaConfig: ReCaptchaConfig
+)(implicit
+    ec: ExecutionContext,
+    clock: Clock
+) {
+  def apply(): Future[Config.Response] = Future {
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/ConfigAction.scala:16]: Given that the value is already computed, `Future.successful { ... }` must be called instead.

File: server/src/main/scala/net/wiringbits/config/ReCaptchaConfig.scala [hunk 10]
```diff
@@ -1,18 +1,19 @@
...
-import net.wiringbits.models.ReCaptchaSecret
+import net.wiringbits.models.{ReCaptchaSecret, SiteKey}
 import play.api.Configuration
 
-case class ReCaptchaConfig(secret: ReCaptchaSecret) {
+case class ReCaptchaConfig(secret: ReCaptchaSecret, siteKey: SiteKey) {
   override def toString: String = {
 
-    s"ReCaptchaConfig(secret = ${secret.toString})"
+    s"ReCaptchaConfig(secret = ${secret.toString}, siteKey = ${siteKey.toString})"
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/config/ReCaptchaConfig.scala:9]: `toString` is unnecessary

File: server/src/main/scala/net/wiringbits/models/SiteKey.scala [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package net.wiringbits.models
+
+import com.typesafe.config.Config
+import play.api.ConfigLoader
+
+case class SiteKey(string: String) extends SecretValue(string)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/models/SiteKey.scala:6]: Let's rename to `RecaptchaSiteKey`, this is not a secrent, hence, `extends SecretValue(string)` is unnecessary

File: server/src/test/scala/controllers/ConfigControllerSpec.scala [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    "return the configuration" in withApiClient { client =>
+      val response = client.getConfig().futureValue
+      response.siteKey must be("6LeIxAcTAAAAAGG")
```
[the remark this answers is not in the document]
Comment (AlexITC) [reply server/src/test/scala/controllers/ConfigControllerSpec.scala:43]: Actually, there is no need to mock it, just create the object in this file and use `inject.bind[ReCaptchaConfig].to(customConfig)`

**Review by AlexITC — CHANGES_REQUESTED** [review]
Nice progress, I left some minor comments.

Thanks.

File: server/src/main/scala/controllers/EnvironmentConfigController.scala [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import scala.concurrent.ExecutionContext
+
+class EnvironmentConfigController @Inject() (
+    configAction: ConfigAction
+)(implicit cc: ControllerComponents, ec: ExecutionContext)
+    extends AbstractController(cc) {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  def getEnvironmentConfig() = handleGET { _ =>
+    logger.info("get frontend config")
```
Comment (caiseL) [inline server/src/main/scala/controllers/EnvironmentConfigController.scala:18]: Capitalize the letter G: `Get frontend config`

File: server/src/main/scala/controllers/EnvironmentConfigController.scala [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    for {
+      response <- configAction()
+      _ = logger.info(response.toString)
```
Comment (caiseL) [inline server/src/main/scala/controllers/EnvironmentConfigController.scala:21]: This logger is unnecessary

File: server/src/main/scala/net/wiringbits/models/SiteKey.scala [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package net.wiringbits.models
+
+import com.typesafe.config.Config
+import play.api.ConfigLoader
+
+case class SiteKey(string: String)
+
+object SiteKey {
```
Comment (caiseL) [inline server/src/main/scala/net/wiringbits/models/SiteKey.scala:8]: Rename it to `ReCaptchaSiteKey` or `RecaptchaSiteKey`

File: server/src/test/scala/controllers/ConfigControllerSpec.scala [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package controllers
+
+import controllers.common.PlayPostgresSpec
+import net.wiringbits.config.ReCaptchaConfig
+
+class ConfigControllerSpec extends PlayPostgresSpec {
+  def environment: ReCaptchaConfig = app.injector.instanceOf(classOf[ReCaptchaConfig])
```
Comment (caiseL) [inline server/src/test/scala/controllers/ConfigControllerSpec.scala:7]: This is not the environment, so rename the variable to `recaptchaConfig`

**Review by caiseL — APPROVED** [review]
Lgtm, only thing missing is loading this environment config on the splash page

File: server/src/main/scala/controllers/EnvironmentConfigController.scala [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+import net.wiringbits.actions.ConfigAction
+import org.slf4j.LoggerFactory
+import play.api.libs.json.Json
+import play.api.mvc.{AbstractController, ControllerComponents}
+
+import javax.inject.Inject
+import scala.concurrent.ExecutionContext
+
+class EnvironmentConfigController @Inject() (
+    configAction: ConfigAction
```
Comment (AlexITC) [inline server/src/main/scala/controllers/EnvironmentConfigController.scala:12]: Let's rename this to `getEnvironmentConfigAction: GetEnvironmentConfigAction`

File: server/src/main/scala/net/wiringbits/actions/ConfigAction.scala [hunk 18]
```diff
@@ -0,0 +1,19 @@
+package net.wiringbits.actions
+
+import net.wiringbits.api.models.GetEnvironmentConfig
+import net.wiringbits.config.ReCaptchaConfig
+
+import java.time.Clock
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class ConfigAction @Inject() (
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/ConfigAction.scala:10]: Rename to `GetEnvironmentConfigAction`

File: server/src/main/scala/net/wiringbits/actions/ConfigAction.scala [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    reCaptchaConfig: ReCaptchaConfig
+)(implicit
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/ConfigAction.scala:12]: No implicit arguments are required

File: server/src/test/scala/controllers/ConfigControllerSpec.scala [hunk 20]
```diff
@@ -0,0 +1,15 @@
+package controllers
+
+import controllers.common.PlayPostgresSpec
+import net.wiringbits.config.ReCaptchaConfig
+
+class ConfigControllerSpec extends PlayPostgresSpec {
```
Comment (AlexITC) [inline server/src/test/scala/controllers/ConfigControllerSpec.scala:6]: Rename to `EnvironmentConfigControllerSpec`, rename the file too

**Review by AlexITC — COMMENTED** [review]
Looks good! I'm happy to merge this once you address the lats comments.

Thanks.

**Review by AlexITC — APPROVED** [review]

**Pull request merged**